index.ts

```diff
@@ -0,0 +1,81 @@
+function isObject(value: unknown): value is Record<string, unknown> {
+  return typeof value === 'object' && value !== null;
+}
+
+type ParametersExceptFirst<F> = F extends (
+  arg0: any,
+  ...rest: infer R
+) => infer ReturnValue
+  ? (...rest: R) => ReturnValue
+  : never;
+type BaseActionType = string;
+type BasePayload = [browser.runtime.MessageSender, ...unknown[]];
+type Method = (...parameters: BasePayload) => Promise<unknown>;
+export type Contract<
+  T extends BaseActionType = BaseActionType,
+  M extends Method = Method,
+  P extends ParametersExceptFirst<M> = ParametersExceptFirst<M>,
+> = {
+  type: T;
+  method: M;
+  publicMethod: P;
+};
+
+const handlers = new Map<BaseActionType, Method>();
+
+export type Message<
+  T extends BaseActionType = BaseActionType,
+  P extends BasePayload = BasePayload,
+> = {
+  type: T;
+  parameters: P;
```

Comment:
Type not call this payload? I think there's two ways to go:

- RPC-style naming, of operationId and args
- Message: type and payload

Reply:
My idea is to only allow RPC-style messaging instead of exposing _two_ interfaces (one for RPC and one for general messages, which isn't necessary if the first one works well).

Names aren't thought out yet, no strong preference since they're going to be on a private object, but since it's RPC-style-only it makes sense to make it an array.

Reply:
I renamed everything to `type` and `args` to keep them short. `type` is the only one visible by the user (I don’t think we'll match the RPC method fully).

I think at some point we'll want to add a library-specific key so we can **safely** catch non-handled messages with it:

```js
browser.runtime.onMessage.addListener(function onMessage(message) {
  if (!message.__messenger__) {
     // Completely ignore message, this was not sent by the library
     // and should not be handled by it
    return;
  }

  const handler = handlers.get(message.type);
  if (!handler) {
    // Even if we don't find the handler, we can safely throw an error,
    // because we're sure it was *supposed* to be handled
    throw new Error('No handler registered for ' + message.type);
  }

  return handler.call(sender, ...message.args);
});
```

Reply:
> My idea is to only allow RPC-style messaging instead of exposing two interfaces

Sure happy to give it a try. In my experience array typing in TypeScript is more annoying that object typing. Arrays also historically are tricky for subtyping: https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Arrays. I'm not familiar with how gracefully TypeScript handles it

The final thing I'll say in support of object payloads is it matches more closely to the browser's sendMessage API

From a functionality standpoint, they're equivalent. E.g., (arg1: Record), {args: unknown[]}. So I'd err on whatever gives the best TypeScript ergonomics.

The benefit of the args based approach I guess is that you can point it to any existing method...

> I think at some point we'll want to add a library-specific key so we can safely catch non-handled messages with it

I agree seems helpful

Reply:
> The final thing I'll say in support of object payloads is it matches more closely to the browser's sendMessage API

I'm confused by this. Arrays are objects and the sendMessage API accepts any serializable value, not just objects. Even primitives are accepted. The actual message is still a `{t:string,a:array}` object.  

The only concern we have is about typing, i.e. ensuring that the handler's type is carried over to the calling point, which is already taken care of.